File: frontend/api_client.py
"""
API Client for Frontend to communicate with Backend API
"""

import requests
from typing import List, Optional, Dict
import os

class APIClient:
    def __init__(self, base_url: Optional[str] = None):
        """
        Initialize API client
        
        Args:
            base_url: Backend API base URL (default: from env or localhost:8000)
        """
        self.base_url = base_url or os.getenv("BACKEND_API_URL", "http://localhost:8000")
    
    def health_check(self) -> Dict:
        """Check if backend API is healthy"""
        try:
            response = requests.get(f"{self.base_url}/health", timeout=5)
            response.raise_for_status()
            return response.json()
        except Exception as e:
            return {"status": "unhealthy", "error": str(e)}
    
    def evaluate_cv(
        self,
        cv_content: str,
        job_description: str,
        custom_rules: str = "",
        llm_model: str = "gemini-2.0-flash"
    ) -> Dict:
        """
        Evaluate CV against job description
        
        Args:
            cv_content: CV text content
            job_description: Job description text
            custom_rules: Custom evaluation rules
            llm_model: LLM model name
            
        Returns:
            Dict with evaluation result
        """
        try:
            response = requests.post(
                f"{self.base_url}/api/evaluate-cv",
                json={
                    "cv_content": cv_content,
                    "job_description": job_description,
                    "custom_rules": custom_rules,
                    "llm_model": llm_model
                },
                timeout=60
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}
    
    def chat(
        self,
        message: str,
        job_description: str,
        custom_rules: str = "",
        cv_evaluations: Optional[List[Dict]] = None,
        chat_history: Optional[List[Dict]] = None,
        llm_model: str = "gemini-2.0-flash"
    ) -> Dict:
        """
        Chat with agent
        
        Args:
            message: User message
            job_description: Job description text
            custom_rules: Custom evaluation rules
            cv_evaluations: List of CV evaluations
            chat_history: Chat history
            llm_model: LLM model name
            
        Returns:
            Dict with chat response
        """
        try:
            response = requests.post(
                f"{self.base_url}/api/chat",
                json={
                    "message": message,
                    "job_description": job_description,
                    "custom_rules": custom_rules,
                    "cv_evaluations": cv_evaluations or [],
                    "chat_history": chat_history or [],
                    "llm_model": llm_model
                },
                timeout=60
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}
    
    # Rules CRUD methods
    def get_all_rules(self) -> Dict:
        """Get all rules"""
        try:
            response = requests.get(f"{self.base_url}/api/rules", timeout=10)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}
    
    def get_rule_names(self) -> Dict:
        """Get all rule names"""
        try:
            response = requests.get(f"{self.base_url}/api/rules/names", timeout=10)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}
    
    def get_rule(self, name: str) -> Dict:
        """Get rule by name"""
        try:
            response = requests.get(f"{self.base_url}/api/rules/{name}", timeout=10)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}
    
    def create_rule(self, name: str, rules: str, description: str = "") -> Dict:
        """Create new rule"""
        try:
            response = requests.post(
                f"{self.base_url}/api/rules",
                json={
                    "name": name,
                    "rules": rules,
                    "description": description
                },
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}
    
    def update_rule(self, name: str, rules: str, description: str = "") -> Dict:
        """Update existing rule"""
        try:
            response = requests.put(
                f"{self.base_url}/api/rules/{name}",
                json={
                    "rules": rules,
                    "description": description
                },
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}
    
    def delete_rule(self, name: str) -> Dict:
        """Delete rule"""
        try:
            response = requests.delete(f"{self.base_url}/api/rules/{name}", timeout=10)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}

File: frontend/src/services/apiClient.ts
import axios from 'axios';
import type { AxiosInstance } from 'axios';

export interface EvaluateCVRequest {
  cv_content: string;
  job_description: string;
  custom_rules?: string;
  llm_model?: string;
}

export interface EvaluateCVResponse {
  success: boolean;
  evaluation?: string;
  error?: string;
}

export interface ChatRequest {
  message: string;
  job_description: string;
  custom_rules?: string;
  cv_evaluations?: Array<{ filename: string; evaluation: string }>;
  chat_history?: Array<{ role: string; content: string }>;
  llm_model?: string;
}

export interface ChatResponse {
  success: boolean;
  response?: string;
  error?: string;
}

export interface Rule {
  name: string;
  rules: string;
  description?: string;
  created_at?: string;
  updated_at?: string;
}

export interface RulesResponse {
  success: boolean;
  rules?: Rule[];
  names?: string[];
  rule?: Rule;
  error?: string;
}

export interface HealthResponse {
  status: string;
  error?: string;
}

class APIClient {
  private client: AxiosInstance;
  private baseURL: string;

  constructor(baseURL?: string) {
    // Use relative URL for production (nginx will proxy to backend)
    // For local development, set VITE_BACKEND_API_URL=http://localhost:8000 in .env
    // Empty string means same origin (relative URLs)
    this.baseURL = baseURL !== undefined ? baseURL : (import.meta.env.VITE_BACKEND_API_URL || '');
    this.client = axios.create({
      baseURL: this.baseURL,
      timeout: 60000,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  async healthCheck(): Promise<HealthResponse> {
    try {
      const response = await this.client.get<HealthResponse>('/health', { timeout: 5000 });
      return response.data;
    } catch (error) {
      return {
        status: 'unhealthy',
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  async evaluateCV(request: EvaluateCVRequest): Promise<EvaluateCVResponse> {
    try {
      const response = await this.client.post<EvaluateCVResponse>('/api/evaluate-cv', {
        cv_content: request.cv_content,
        job_description: request.job_description,
        custom_rules: request.custom_rules || '',
        llm_model: request.llm_model || 'gemini-2.0-flash',
      });
      return response.data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  async chat(request: ChatRequest): Promise<ChatResponse> {
    try {
      const response = await this.client.post<ChatResponse>('/api/chat', {
        message: request.message,
        job_description: request.job_description,
        custom_rules: request.custom_rules || '',
        cv_evaluations: request.cv_evaluations || [],
        chat_history: request.chat_history || [],
        llm_model: request.llm_model || 'gemini-2.0-flash',
      });
      return response.data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  async getAllRules(): Promise<RulesResponse> {
    try {
      const response = await this.client.get<RulesResponse>('/api/rules', { timeout: 10000 });
      return response.data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  async getRuleNames(): Promise<RulesResponse> {
    try {
      const response = await this.client.get<RulesResponse>('/api/rules/names', { timeout: 10000 });
      return response.data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  async getRule(name: string): Promise<RulesResponse> {
    try {
      const response = await this.client.get<RulesResponse>(`/api/rules/${name}`, { timeout: 10000 });
      return response.data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  async createRule(name: string, rules: string, description?: string): Promise<RulesResponse> {
    try {
      const response = await this.client.post<RulesResponse>('/api/rules', {
        name,
        rules,
        description: description || '',
      });
      return response.data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  async updateRule(name: string, rules: string, description?: string): Promise<RulesResponse> {
    try {
      const response = await this.client.put<RulesResponse>(`/api/rules/${name}`, {
        rules,
        description: description || '',
      });
      return response.data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  async deleteRule(name: string): Promise<RulesResponse> {
    try {
      const response = await this.client.delete<RulesResponse>(`/api/rules/${name}`);
      return response.data;
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }
}

export const apiClient = new APIClient();
export default APIClient;

